### PR TITLE
delete draft application mutation

### LINKED ISSUE #3416: API mutation to delete draft applications
## Description

Create a backend mutation which deletes an application (aka PoolCandidate object). This must be protected by a Policy which requires the user be the owner of the Application (or an admin) and that the application is still a draft.

QUESTION: Do we want to set up proper error codes for this?
ANSWER: This is a destructive action. Probably best to do all the things.

## Acceptance Criteria
- [x] mutation deletes a PoolCandidate specified by ID
- [x] mutation fails if user doesn't own the application and isn't an admin.
- [x] mutation fails if application is not a draft
- [x] if mutation fails, it should return an error code which explains which reason it failed for
- [x] Hard delete. But talk to Gray if soft delete is much easier.
- [ ] Create a frontend message for these two error codes ???

### PR DESCRIPTION
**Pull request by vd1992**
resolves #3416 

adds application deletion via hard delete
the bit marked with "??" in the issue I am unsure of whether it was desired yet as frontend error messages are still a work in progress

test: use this mutation
```
mutation deleteTest {
  deleteApplication(id: "APPLICATION ID") 
}
```

### CODE REVIEW

**Review by esizer — CHANGES_REQUESTED** [review]
Looking great! One small naming change and a question :)

File: api/app/GraphQL/Mutations/DeleteApplication.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * Delete an application
+     * @param  null  $_
+     * @param  array{}  $args
+     */
+    public function __invoke($_, array $args)
+    {
+        // grab the specific application, validate to ensure the status of the application is DRAFT or DRAFT_EXPIRED
+        $application = PoolCandidate::find($args['id']);
+        $archivalValidator = new DeleteApplicationValidator;
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Mutations/DeleteApplication.php]: oops! fail!
fixed here https://github.com/GCTC-NTGC/gc-digital-talent/pull/3760/commits/bdfd8361ff0c4b57f00cedeacdf1d9fde8afa9a7

File: api/app/GraphQL/Mutations/DeleteApplication.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        $validator = Validator::make($application->toArray(), $archivalValidator->rules(), $archivalValidator->messages());
+        if ($validator->fails()) {
+            throw new ValidationException($validator->errors()->first(), $validator);
+        }
+
+        // clear connected tables
+        $application->cmoAssets()->sync([]);
+        $application->expectedClassifications()->sync([]);
+        $application->save();
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Mutations/DeleteApplication.php:29]: no idea
without the clear connected tables bit the resolver failed on the account of the id of the being deleted item still being referenced elsewhere

[the remark this answers is not in the document]
Comment (esizer) [reply api/app/GraphQL/Mutations/DeleteApplication.php:29]: Yeah, I'm not sure how we set up the schema. This smells of us not defining `onDelete` logic in the migration. Not a big deal just results in us needing to write more precise and careful code :)

**Review by esizer — APPROVED** [review]
Awesome! 😀

File: api/app/GraphQL/Mutations/DeleteApplication.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Mutations/DeleteApplication.php:29]: I see, thanks for the advice!

**Pull request merged**